packages/backend/src/core/PerpetualRollupUpdater.ts

```diff
@@ -0,0 +1,99 @@
+import { OnChainData } from '@explorer/encoding'
+import {
+  calculateUpdatedPositions,
+  IMerkleStorage,
+  MerkleTree,
+  PositionLeaf,
+} from '@explorer/state'
+import { PedersenHash } from '@explorer/types'
+
+import { ForcedTransactionsRepository } from '../peripherals/database/ForcedTransactionsRepository'
+import { PageRepository } from '../peripherals/database/PageRepository'
+import { StateTransitionRecord } from '../peripherals/database/StateTransitionRepository'
+import { StateUpdateRepository } from '../peripherals/database/StateUpdateRepository'
+import { EthereumClient } from '../peripherals/ethereum/EthereumClient'
+import { Logger } from '../tools/Logger'
+import { StateUpdater } from './StateUpdater'
+
+/**
+ * @internal
+ * Same as `await RollupState.empty().then(empty => empty.positions.hash())`
+ */
+export const ROLLUP_STATE_EMPTY_HASH = PedersenHash(
+  '52ddcbdd431a044cf838a71d194248640210b316d7b1a568997ecad9dec9626'
+)
+
+const positionTreeHeight = 64n
+
+export class PerpetualRollupUpdater extends StateUpdater<PositionLeaf> {
+  constructor(
+    private readonly pageRepository: PageRepository,
+    protected readonly stateUpdateRepository: StateUpdateRepository,
+    protected readonly rollupStateRepository: IMerkleStorage<PositionLeaf>,
```

Comment:
rollupStateRepository -> merkleStorage.

Also maybe IMerkleStorage should be renamed to IMerkleTreeRepository?